src/Data/Variant.hs

```diff
@@ -276,6 +279,36 @@ class CouldBeF xs x => CouldBe (xs :: [Type]) (x :: Type) where
 instance CouldBeF xs x => CouldBe xs x where
   throw = throwF . Identity
 
+type family FlattenConstraints (cs :: [Constraint]) = (c :: Constraint) | c -> cs where
+  FlattenConstraints  '[] = ()
+  FlattenConstraints (c ': cs) = (c, FlattenConstraints cs)
+
+type family TypeMap (f :: k -> l) (xs :: [k]) = (c :: [l]) where
+  TypeMap f (x ': xs) = f x ': (TypeMap f xs)
+  TypeMap f '[] = '[]
+
+type family SpreadConstraint (f :: k -> Constraint) (xs :: [k]) = (c :: Constraint) where
+  SpreadConstraint f xs = FlattenConstraints (TypeMap f xs)
+
+-- | As with 'CouldBeAnyOf', we can also constrain a variant to represent
+-- several possible types, as we might with several 'CouldBeF' constraints,
+-- using one type-level list.
+type family CouldBeAnyOfF (e :: [Type]) (xs :: [Type]) = (c :: Constraint) where
+  CouldBeAnyOfF e xs = SpreadConstraint (CouldBeF e) xs
+
+-- | Listing larger variants' constraints might amplify the noise of
+-- functions' signatures. The 'CouldBeAnyOfF' constraint lets us specify
+-- several types a variant may contain in a single type level list, as opposed
+-- to several independent constraints. So, we could replace,
+--
+-- f :: (e `CouldBe` Int, e `CouldBe` Bool, e `CouldBe` Char) => VariantF IO e
```

Comment:
The comments are actually checked with [`doctest`](http://hackage.haskell.org/package/doctest), so it might be nice to throw in a couple (`>>>`-prefixed) examples of each, just to show that `catch` does indeed still cancel out `throw`!